### PR TITLE
AGDR/ Configure Overview section

### PR DESCRIPTION
**Pull request by eadams17**
## WHAT
configure Overview section of the Admin Growth Diagnostic Reports to aggregate and display data from the backend

## WHY
now that the necessary APIs are available, we want to configure them to feed into the frontend UI

## HOW
- add a bunch of logic for making the various API calls and then aggregating the data once they are all receive
- update styling of the table

**just a note that there are quite a few snapshots that were updated-- this is because I moved the `DataTable` header name inside of a <span> element for more styling flexibility**

### Screenshots
(If applicable. Also, please censor any sensitive data)
Screen Shot 2023-12-05 at 2 49 46 PM
Screen Shot 2023-12-05 at 2 49 58 PM
Screen Shot 2023-12-05 at 2 51 55 PM
Screen Shot 2023-12-05 at 2 52 25 PM
Screen Shot 2023-12-05 at 2 52 57 PM

### Notion Card Links
(Please provide links to any relevant Notion card(s) relevant to this PR.)
https://www.notion.so/quill/Admin-Diagnostic-Growth-Report-Overview-Frontend-8e22794d03f345af85b1d95a62e77098?pvs=4

PR Checklist | Your Answer
------------ | -------------
Have you added and/or updated tests? |  yes
Have you deployed to Staging? | yes
Self-Review: Have you done an initial self-review of the code below on Github? | yes
Spec Review: Have you reviewed the spec and ensured this meets requirements and/or matches design mockups? | yes

### CODE REVIEW

File: services/QuillLMS/app/workers/admin_diagnostic_reports/diagnostic_overview_worker.rb [hunk 1]
```diff
@@ -4,7 +4,7 @@ module AdminDiagnosticReports
   class DiagnosticOverviewWorker
     include Sidekiq::Worker
 
-    PUSHER_EVENT = 'admin-diagnostic-ovewview-cached'
+    PUSHER_EVENT = 'admin-diagnostic-overview-cached'
```
Comment (eadams17) [inline services/QuillLMS/app/workers/admin_diagnostic_reports/diagnostic_overview_worker.rb:7]: @anathomical I noticed this misspelling-- is there anywhere else I need to worry about updating this value? (I already updated the `PUSHER_EVENT_KEY` in the `OverviewSection` frontend file and didn't find anywhere else when searching the codebase, but I wanted to check with you just to be sure)

File: services/QuillLMS/client/app/bundles/PremiumHub/containers/__tests__/overviewSection.test.tsx [hunk 2]
```diff
@@ -34,22 +37,21 @@ describe('OverviewSection', () => {
...
+  //     expect(screen.getByRole('columnheader', { name: /average activities & time spent/i })).toBeInTheDocument()
+  //     expect(screen.getByRole('columnheader', { name: /post diagnostic completed/i })).toBeInTheDocument()
+  //     expect(screen.getByRole('columnheader', { name: /overall skill growth/i })).toBeInTheDocument()
+  //   })
+  //   test('it should render the expected empty state message', () => {
+  //     const { asFragment } = render(<OverviewSection {...props} />);
+  //     expect(asFragment()).toMatchSnapshot();
+  //     expect(screen.getByText(/there is no diagnostic data available for the filters selected\. try modifying or removing a filter to see results\./i)).toBeInTheDocument()
+  //   })
+  // })
```
Comment (eadams17) [inline services/QuillLMS/client/app/bundles/PremiumHub/containers/__tests__/overviewSection.test.tsx:58]: I was trying to use the `passedData` convention we use in some other test files to stub the incoming data(i.e. [https://github.com/empirical-org/Empirical-Core/blob/07fc91f79f408045129e496356cb61dddfb2842c/services/QuillLMS/client/app/bundles/PremiumHub/components/canvas/container.tsx#L13](url)), but I kept hitting some issues. Since this is time sensitive and I want to give Peter S the chance to start testing ASAP, I opted to punt on adding more thorough testing in a later PR.

File: services/QuillLMS/app/assets/stylesheets/pages/premium_hub/filterable-reports.scss [hunk 3]
```diff
@@ -41,6 +41,19 @@
...
+                .dropdown__value-container {
+                  .dropdown__single-value {
+                    color: #7f7f7f;
+                  }
+                }
+              }
+            }
+          }
+        }
     }
```
Comment (emilia-friedberg) [inline services/QuillLMS/app/assets/stylesheets/pages/premium_hub/filterable-reports.scss]: I think I would just write this as `.timeframe-filter.disabled .dropdown .dropdown__control .dropdown__value-container .dropdown__single-value { color: #7f7f7f }` to avoid all the nesting. Also, should that color be a variable?

File: services/QuillLMS/client/app/bundles/PremiumHub/containers/__tests__/overviewSection.test.tsx (hunk 2 above)
Comment (emilia-friedberg) [reply services/QuillLMS/client/app/bundles/PremiumHub/containers/__tests__/overviewSection.test.tsx:58]: I'm okay with it not being in this PR, but I think it should probably be the next one (can you work on the tests while Peter Sharkey does his review?)

File: services/QuillLMS/client/app/bundles/PremiumHub/containers/diagnosticGrowthReports/helpers.tsx [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+export const averageActivitiesAndTimeSpentTooltipText = "Each diagnostic is linked to recommended practice activities. For the students who have completed activities, this shows the average number of completed activities and the average time spent per student.<br/><br/> This counts the practice activities connected to this particular diagnostic - not the total number of activities that the student has practiced on Quill."
+export const postDiagnosticCompletedTooltipText = "The total number of students who completed the Post diagnostic of all of the students assigned the Post diagnostic.<br/><br/> Students are not included in this count until their teacher assigns the Post diagnostic to them."
+export const overallSkillGrowthTooltipText = "The average increase in growth scores across all of the skills.<br/><br/> The Performance by Skill report shows the average increase in questions answered correctly for each skill, and the overall growth is the average increase across all skills."
+
+export function aggregateOverviewData({
```
Comment (emilia-friedberg) [inline services/QuillLMS/client/app/bundles/PremiumHub/containers/diagnosticGrowthReports/helpers.tsx]: This function is really hard for me to parse. I think we need to split out some interior functions and work on formatting for readability. I think this is a good use case for getting help from ChatGPT -- here's an initial stab at it that might be useful: https://chat.openai.com/c/97307d69-7fcd-4b56-9c38-f8e884c5be4d. I think we also probably want some tests specific to this method (the same way we would about this object if it were being sent up from a controller).

File: services/QuillLMS/client/app/bundles/PremiumHub/containers/diagnosticGrowthReports/index.tsx [hunk 5]
```diff
@@ -28,17 +28,78 @@ export const DiagnosticGrowthReportsContainer = ({
...
+  }
+
+  function renderContent() {
+
+    if (noDiagnosticDataAvailable) {
+      return(
+        <div className="no-diagnostic-data-container">
+          <h2>There are not yet any completed diagnostics.</h2>
+          <p>The Quill Diagnostic enables you to provide a Pre-Diagnostic at the start of the year to identify skills students need to practice. The Post-Diagnostic allows you to then measure students&apos; learning gains over the course of the school year.</p>
+          <p>At the moment, there are no teachers connected to your admin account who have completed a diagnostic. Once at least one teacher has completed a diagnostic with at least one student, you will be able to see their results in this report. View our guide our guide to learn how teachers can assign diagnostics to their students.</p>
```
Comment (emilia-friedberg) [inline services/QuillLMS/client/app/bundles/PremiumHub/containers/diagnosticGrowthReports/index.tsx]: "our guide" is duplicated in the last sentence on this line.

File: services/QuillLMS/client/app/bundles/PremiumHub/containers/diagnosticGrowthReports/index.tsx [hunk 6]
```diff
@@ -28,17 +28,78 @@ export const DiagnosticGrowthReportsContainer = ({
...
+            <span>Performance Overview</span>
+          </button>
+          <button className={`interactive-wrapper performance-type-button skill ${activeTab === SKILL ? 'active' : ''}`} onClick={handleTabChange} value={SKILL}>
+            <img alt="" src={activeTab === SKILL ? pencilWhiteIconSrc : pencilGreyIconSrc} />
+            <span>Performance by Skill</span>
+          </button>
+          <button className={`interactive-wrapper performance-type-button student ${activeTab === STUDENT ? 'active' : ''}`} onClick={handleTabChange} value={STUDENT}>
+            <img alt="" src={activeTab === STUDENT ? groupOfStudentsWhiteIconSrc : groupOfStudentsGreyIconSrc} />
+            <span>Performance by Student</span>
+          </button>
```
Comment (emilia-friedberg) [inline services/QuillLMS/client/app/bundles/PremiumHub/containers/diagnosticGrowthReports/index.tsx]: All the ternary logic in here is effectively duplicated and feels a little messy. I think we should do something like:

```
const performanceButtons = [
  {
    tab: OVERVIEW,
    displayName: 'Performance Overview',
    inactiveIconSrc: barChartGreySrc,
    activeIconSrc: barChartWhiteIconSrc
  },
  ...
]
```

Then iterate through each of them to generate the button elements in their own function.

File: services/QuillLMS/client/app/bundles/PremiumHub/containers/diagnosticGrowthReports/overviewSection.tsx [hunk 7]
```diff
@@ -1,96 +1,256 @@
...
+    setLoading(true)
+    getPreDiagnosticAssignedData()
+    getPreDiagnosticCompletedData()
+    getPostDiagnosticAssignedData()
+    getPostDiagnosticCompletedData()
+    getRecommendationsData()
+  }
+
+  function getPreDiagnosticAssignedData () {
+    const searchParams = getSearchParams("pre-diagnostic-assigned")
```
Comment (emilia-friedberg) [inline services/QuillLMS/client/app/bundles/PremiumHub/containers/diagnosticGrowthReports/overviewSection.tsx]: Should these querystrings be constants?

File: services/QuillLMS/app/workers/admin_diagnostic_reports/diagnostic_overview_worker.rb (hunk 1 above)
Comment (anathomical) [reply services/QuillLMS/app/workers/admin_diagnostic_reports/diagnostic_overview_worker.rb:7]: @eadams17 Good catch.  I think those should be the only two places where that value is defined: once on the front-end and once on the back-end.

File: services/QuillLMS/client/app/bundles/PremiumHub/containers/__tests__/overviewSection.test.tsx (hunk 2 above)
Comment (anathomical) [reply services/QuillLMS/client/app/bundles/PremiumHub/containers/__tests__/overviewSection.test.tsx:58]: If we're going to leave this commented out after a merge, we should probably add a `// TODO: ` note to make it easier to find this later.

File: services/QuillLMS/client/app/bundles/PremiumHub/shared.tsx [hunk 8]
```diff
@@ -80,3 +81,15 @@ export function hashPayload(payloadArray: Array<any>) {
...
 }
+
+export function getTimeInMinutesAndSeconds(seconds) {
+  if(!seconds) return NOT_APPLICABLE
+
+  let numminutes = Math.floor((((seconds % 31536000) % 86400) % 3600) / 60).toString();
+  let numseconds = Math.floor((((seconds % 31536000) % 86400) % 3600) % 60).toString();
+
+  if (numminutes.length === 1) numminutes = "0" + numminutes
+  if (numseconds.length === 1) numseconds = "0" + numseconds
```
Comment (anathomical) [inline services/QuillLMS/client/app/bundles/PremiumHub/shared.tsx]: I think you can use `toString().padLeft(2, '0')` instead of adding these `if length === 1` conditions.

File: services/QuillLMS/client/app/bundles/PremiumHub/containers/diagnosticGrowthReports/skillSection.tsx [hunk 9]
```diff
@@ -77,21 +77,13 @@ const headers = [
...
-  selectedTeachers,
-  availableTeachers,
+  selectedSchoolIds,
+  selectedTeacherIds,
+  selectedClassroomIds,
   selectedTimeframe,
-  handleClickDownloadReport,
-  openMobileFilterMenu
+  pusherChannel
 }) => {
```
Comment (emilia-friedberg) [inline services/QuillLMS/client/app/bundles/PremiumHub/containers/diagnosticGrowthReports/skillSection.tsx:87]: Love the clean up here.

File: services/QuillLMS/client/app/bundles/PremiumHub/shared.tsx [hunk 10]
```diff
@@ -80,3 +81,15 @@ export function hashPayload(payloadArray: Array<any>) {
   const joinedPayload = payloadArray.join('-')
   return md5(joinedPayload)
 }
+
+export function getTimeInMinutesAndSeconds(seconds) {
+  if(!seconds) return NOT_APPLICABLE
+
+  let numminutes = Math.floor((((seconds % 31536000) % 86400) % 3600) / 60).toString();
+  let numseconds = Math.floor((((seconds % 31536000) % 86400) % 3600) % 60).toString();
```
Comment (emilia-friedberg) [inline services/QuillLMS/client/app/bundles/PremiumHub/shared.tsx]: It would be good to have these numbers as constants so it's clear what they mean.

**Review by emilia-friedberg — CHANGES_REQUESTED** [review]
There's a lot of really good work in here -- just a couple of places I think we need to clean things up before they're ready to go to production.

File: services/QuillLMS/client/app/bundles/PremiumHub/containers/diagnosticGrowthReports/helpers.tsx [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+    const preDiagnosticScore = preDiagnosticCompletedDataHash[id]?.pre_average_score
+    const postStudentsAssigned = postDiagnosticAssignedDataHash[id]?.post_students_assigned
+    const postStudentsCompleted = postDiagnosticCompletedDataHash[id]?.post_students_completed
+    const postDiagnosticScore = postDiagnosticCompletedDataHash[id]?.post_average_score
+    const studentsCompletedPractice = recommendationsDataHash[id]?.students_completed_practice
+    const averageActivitiesCount = recommendationsDataHash[id]?.average_practice_activities_count
+    const averageTimespent = recommendationsDataHash[id]?.average_time_spent_seconds
+    const aggregateRowsDataForDiagnostic = aggregateRowsData[id]
+    let overallSkillGrowth: string | JSX.Element = 'No Growth'
+    if (!preDiagnosticScore || !postDiagnosticScore) {
```
Comment (anathomical) [inline services/QuillLMS/client/app/bundles/PremiumHub/containers/diagnosticGrowthReports/helpers.tsx]: I think this conditional is fine as-is, but as far as I know it should be impossible for a row to have a `postDiagnosticScore` but not a `preDiagnosticScore`, so I think practically speaking you can just make this `if (!postDiagnosticScore)` if you wanted, but like I said, it's fine as-is.

File: services/QuillLMS/client/app/bundles/PremiumHub/containers/diagnosticGrowthReports/overviewSection.tsx [hunk 12]
```diff
@@ -1,96 +1,256 @@
...
+import { groupByDropdownOptions, hashPayload } from '../../shared'
+import { requestPost, } from '../../../../modules/request';
+import { aggregateOverviewData, averageActivitiesAndTimeSpentTooltipText, completedActivitiesTooltipText, diagnosticNameTooltipText, overallSkillGrowthTooltipText, postDiagnosticCompletedTooltipText, preDiagnosticCompletedTooltipText } from './helpers';
+
+const QUERY_KEY = "admin-diagnostic-overview"
+const PUSHER_EVENT_KEY = "admin-diagnostic-overview-cached";
+const DIAGNOSTIC_NAME = 'Diagnostic Name'
+const PRE_DIAGNOSTIC_COMPLETED = 'Pre Diagnostic Completed'
+const COMPLETED_ACTIVITIES = 'Completed Activities'
+const OVERALL_SKILL_GROWTH = 'Overall Skill Growth'
```
Comment (anathomical) [inline services/QuillLMS/client/app/bundles/PremiumHub/containers/diagnosticGrowthReports/overviewSection.tsx]: Totally dig these as constants.

File: services/QuillLMS/client/app/bundles/PremiumHub/containers/diagnosticGrowthReports/overviewSection.tsx [hunk 13]
```diff
@@ -1,96 +1,256 @@
...
 
 const headers = [
   {
-    name: 'Diagnostic Name',
+    name: DIAGNOSTIC_NAME,
     attribute: 'name',
-    width: DIAGNOSTIC_REPORT_DEFAULT_CELL_WIDTH,
-    tooltipName: '',
-    tooltipDescription: '',
+    width: '176px',
```
Comment (anathomical) [inline services/QuillLMS/client/app/bundles/PremiumHub/containers/diagnosticGrowthReports/overviewSection.tsx]: Given that all of these widths seem to be the same value, I wonder if this shouldn't also be a constant even if it's not from a shared file??

File: services/QuillLMS/client/app/bundles/PremiumHub/containers/diagnosticGrowthReports/overviewSection.tsx [hunk 14]
```diff
@@ -1,96 +1,256 @@
...
+  }
+
+  function filtersMatchHash(hashMessage) {
+    const filterTarget = [].concat(
+      QUERY_KEY,
+      selectedTimeframe,
+      selectedSchoolIds,
+      selectedGrades,
+      selectedTeacherIds,
+      selectedClassroomIds
```
Comment (anathomical) [inline services/QuillLMS/client/app/bundles/PremiumHub/containers/diagnosticGrowthReports/overviewSection.tsx:237]: So this isn't currently in the back-end, but reviewing this suggests to me that we should also have `groupByValue` in here.  Let's just do that in a separate PR, though, to avoid adding complexity to this one.

**Review by anathomical — APPROVED** [review]
A couple of minor notes, but overall this looks good to me.

File: services/QuillLMS/app/assets/stylesheets/pages/premium_hub/filterable-reports.scss (hunk 3 above)
Comment (eadams17) [reply services/QuillLMS/app/assets/stylesheets/pages/premium_hub/filterable-reports.scss]: good call! I'll update. I already flagged the hex code to Jack so that he can map it to a variable so just waiting for that (it's used quite a bit around the codebase)

File: services/QuillLMS/client/app/bundles/PremiumHub/containers/diagnosticGrowthReports/index.tsx (hunk 6 above)
Comment (eadams17) [reply services/QuillLMS/client/app/bundles/PremiumHub/containers/diagnosticGrowthReports/index.tsx]: love that suggestion-- I'll make this update

File: services/QuillLMS/client/app/bundles/PremiumHub/containers/diagnosticGrowthReports/overviewSection.tsx (hunk 7 above)
Comment (eadams17) [reply services/QuillLMS/client/app/bundles/PremiumHub/containers/diagnosticGrowthReports/overviewSection.tsx]: yeah, I can make these constants

File: services/QuillLMS/client/app/bundles/PremiumHub/containers/diagnosticGrowthReports/helpers.tsx (hunk 11 above)
Comment (eadams17) [reply services/QuillLMS/client/app/bundles/PremiumHub/containers/diagnosticGrowthReports/helpers.tsx]: that makes sense-- I'll update to be more succinct

File: services/QuillLMS/client/app/bundles/PremiumHub/containers/diagnosticGrowthReports/overviewSection.tsx (hunk 14 above)
Comment (eadams17) [reply services/QuillLMS/client/app/bundles/PremiumHub/containers/diagnosticGrowthReports/overviewSection.tsx:237]: sounds good to me!

File: services/QuillLMS/client/app/bundles/PremiumHub/containers/diagnosticGrowthReports/helpers.tsx (hunk 4 above)
Comment (eadams17) [reply services/QuillLMS/client/app/bundles/PremiumHub/containers/diagnosticGrowthReports/helpers.tsx]: @emilia-friedberg I can't seem to access the link-- can you maybe send the prompt you inputted so I can have better context?

File: services/QuillLMS/client/app/bundles/PremiumHub/containers/__tests__/diagnosticGrowthReportsHelpers.test.tsx [hunk 15]
```diff
@@ -0,0 +1,214 @@
...
+      mockArgs.preDiagnosticCompletedData = mockPreDiagnosticCompletedData
+      mockArgs.postDiagnosticAssignedData = mockPostDiagnosticAssignedData
+      mockArgs.postDiagnosticCompletedData = mockPostDiagnosticCompletedData
+      mockArgs.recommendationsData = mockRecommendationsData
+      aggregateOverviewData(mockArgs)
+      expect(mockArgs.setLoading).toHaveBeenCalledWith(false)
+      expect(mockArgs.setAggregatedData).toHaveBeenCalledWith(mockCombinedData)
+    })
+  })
+})
```
Comment (emilia-friedberg) [inline services/QuillLMS/client/app/bundles/PremiumHub/containers/__tests__/diagnosticGrowthReportsHelpers.test.tsx:214]: love these tests!

File: services/QuillLMS/client/app/bundles/PremiumHub/containers/diagnosticGrowthReports/index.tsx [hunk 16]
```diff
@@ -58,35 +71,44 @@ export const DiagnosticGrowthReportsContainer = ({
...
+  }
+
+  function renderButtons() {
+    return(
+      <div className="tabs-for-pages-container">
+        {reportButtons.map(button => {
+          const { tab, displayName, activeIconSrc, inactiveIconSrc } = button
+          return (
+            <button className={`interactive-wrapper performance-type-button overview ${tab === OVERVIEW ? 'active' : ''}`} onClick={handleTabChange} value={tab}>
+              <img alt="" src={tab === OVERVIEW ? activeIconSrc : inactiveIconSrc} />
```
Comment (emilia-friedberg) [inline services/QuillLMS/client/app/bundles/PremiumHub/containers/diagnosticGrowthReports/index.tsx]: should these two have `activeTab === tab` as the condition instead of `tab === OVERVIEW`?

**Review by emilia-friedberg — APPROVED** [review]
The refactor you did with that function is so much easier to parse (and will be easier to debug if we ever need to now) -- beautiful work! Love the tests, too. I flagged one small thing I think may have been an error in the refactor but I don't need to review again. Thank you so much for doing all that. Excited to see this ship!

File: services/QuillLMS/client/app/bundles/PremiumHub/containers/diagnosticGrowthReports/index.tsx (hunk 16 above)
Comment (eadams17) [reply services/QuillLMS/client/app/bundles/PremiumHub/containers/diagnosticGrowthReports/index.tsx]: ahh yeah, good catch! I had resolved it in my other branch but not here
